engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java

```diff
@@ -140,16 +150,22 @@ public void unsubscribeFromEvents(
     return expressionProcessor
         .evaluateMessageCorrelationKeyExpression(expression, scopeKey)
         .map(BufferUtil::wrapString)
+        .map(evaluation::recordCorrelationKey)
         .mapLeft(f -> new Failure(f.getMessage(), f.getErrorType(), scopeKey));
   }
 
-  private Either<Failure, Timer> evaluateTimer(
-      final ExecutableCatchEvent event, final BpmnElementContext context) {
+  private Either<Failure, OngoingEvaluation> evaluateTimer(final OngoingEvaluation evaluation) {
```

Comment:
Played around with it a little bit. Please have a look and let me know what you think.

I felt extracting the evaluation logic was a good step. Now there are two medium classes instead of one big one. The other steps, not so sure about. I basically kept going refactoring so it would be easier - in my mind - to understand the code. 

I think testability was improved. Still not happy with the code, though. Maybe pattern matching would come in handy, once it is available in Java :thinking: 

The thing that bothers me is that in some parts of the code we treat timer and message events the same, but in other codes it is clear that it makes only sense for either of them..

Anyway, just have a look and maybe you get to new ideas.

Reply:
I've had some time to review the code and play around with it locally. I've written down my thoughts in the review below. Note that a bunch of the comments are personal opinions. I hope that's okay since you asked for it specifically.